replica/database.cc

```diff
@@ -1723,10 +1723,31 @@ future<mutation> database::apply_counter_update(schema_ptr s, const frozen_mutat
   }));
 }
 
+// #9919 etc. The initiative to wrap exceptions here
+// causes a bunch of problems with (implicit) call sites
+// catching timed_out_error (not checking is_timeout_exception).
+// Fixing the call sites is a good idea, but it is also hard
+// to verify. This workaround should ensure we take the
+// correct code paths in all cases, until we can clean things up
+// proper.
+class wrapped_timed_out_error : public timed_out_error {
+private:
+    sstring _msg;
+public:
+    wrapped_timed_out_error(sstring msg)
+        : _msg(std::move(msg))
+    {}
+    const char* what() const noexcept override {
+        return _msg.c_str();
+    }
+};
+
+// see above (#9919)
+template<typename T = std::runtime_error>
 static void throw_commitlog_add_error(schema_ptr s, const frozen_mutation& m) {
     // it is tempting to do a full pretty print here, but the mutation is likely
     // humungous if we got an error, so just tell us where and pk...
-    std::throw_with_nested(std::runtime_error(format("Could not write mutation {}:{} ({}) to commitlog"
+    std::throw_with_nested(T(format("Could not write mutation {}:{} ({}) to commitlog"
```

Comment:
What's the point of `throw_with_nested`?
Why is it needed?

Reply:
The point is we want to know the table and pk, tracing back to older issues with errors writing but no log info on who/what wrote. We could log here, but it is cleaner to allow caller to either log or handle things. Unfortunately we broke some catch points.

Reply:
But the table and pk are printed into the exception message, why do need to throw it as a nested exception?